Framework/src/BookkeepingQualitySink.cxx

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file   BookkeepingQualitySink.cxx
+/// \author Michal Tichak
+///
+
+#include "QualityControl/BookkeepingQualitySink.h"
+#include <Framework/DataRefUtils.h>
+#include <Framework/InputRecordWalker.h>
+#include "QualityControl/Bookkeeping.h"
+#include "QualityControl/QualitiesToFlagCollectionConverter.h"
+#include "QualityControl/QualityObject.h"
+#include "QualityControl/QcInfoLogger.h"
+#include <BookkeepingApi/QcFlagServiceClient.h>
+#include <BookkeepingApi/BkpClientFactory.h>
+#include <stdexcept>
+
+namespace o2::quality_control::core
+{
+
+void BookkeepingQualitySink::send(const std::string& grpcUri, const BookkeepingQualitySink::FlagsMap& flags, Provenance type)
+{
+  auto bkpClient = o2::bkp::api::BkpClientFactory::create(grpcUri);
+  auto& qcClient = bkpClient->qcFlag();
+
+  ILOG(Info, Support) << "Sending " << flags.size() << " flag collections" << ENDM;
+
+  for (const auto& [_, flagCollection] : flags) {
+
+    std::vector<QcFlag> bkpQcFlags{};
+    for (const auto& flag : *flagCollection) {
+      // BKP uses start/end of run for missing time values, so we are using this functionality in order to avoid
+      // determining these values by ourselves (see TaskRunner::start() for details). mtichak checked with mboulais that
+      // it is okay to do so.
+      bkpQcFlags.emplace_back(QcFlag{
+        .flagTypeId = flag.getFlag().getID(),
+        .from = flag.getStart() == gFullValidityInterval.getMin() ? std::nullopt : std::optional<uint64_t>{ flag.getStart() },
+        .to = flag.getEnd() == gFullValidityInterval.getMax() ? std::nullopt : std::optional<uint64_t>{ flag.getEnd() },
+        .origin = flag.getSource(),
+        .comment = flag.getComment() });
+    }
+
+    if (bkpQcFlags.empty()) {
+      continue;
+    }
```

Comment:
Since `bkpQcFlags` will basically have the same number of elements as `*flagCollection`, could you do input validation earlier, i.e. continue if `flagCollection->empty()` at the beginning of the loop?